cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,323 @@
+
+				prevtranscribedSeg := -1
+				for i, seg := range segments {
+					if prevTranscribedPos >= seg.Start && prevTranscribedPos < seg.End {
+						prevtranscribedSeg = i
+						break
+					}
+				}
+
+				if prevtranscribedSeg >= 0 {
+					allSilence := true
+					for i := prevtranscribedSeg; i < len(segments); i++ {
+						if !segments[i].Silence {
+							allSilence = false
+							break
+						}
+					}
+					if allSilence {
+						silenceLength := segments[len(segments)-1].End - segments[prevtranscribedSeg].Start
+						if silenceLength >= int(removeWindowSilenceInSamples) {
+							// 1. untranscribed data is all silence, and there's been enough silence to end this window.
+							fmt.Printf("segLen: %d, remove after: %d, clearing window!\n", silenceLength, removeWindowSilenceInSamples)
+							window = window[:0]
+							prevTranscribedPos = 0
+							return
+						}
+						// 2. all new (untranscribed) data is silence, so don't send to transcriber.
+						fmt.Printf("-- all untranscribed data is silence ** not sending to transcriber\n")
+						return
+					}
+				}
+
+				// Track our new position and send off data for transcription
+				prevTranscribedPos = len(cleaned)
+				transcribed := make(chan string)
+				t.captionQueue <- captionPackage{
+					pcm: cleaned,
+					ret: transcribed,
+				}
+
+				for {
+					select {
+					case <-ticker.C:
+						// TODO: add metrics for this.
+						slog.Debug("live captions: dropped a tick waiting for the transcriber",
+							slog.String("trackID", ctx.trackID))
+					case text := <-transcribed:
+						if err := t.client.Send(wsEvPrefix+"caption", CaptionMsg{
+							SessionID: ctx.sessionID,
+							UserID:    ctx.user.Id,
+							Text:      text,
+						}, false); err != nil {
+							slog.Error("live captions: error sending ws captions",
+								slog.String("err", err.Error()),
+								slog.String("trackID", ctx.trackID))
+						}
+
+						return
+					}
+				}
+			}()
+		}
+	}
+}
+
+func (t *Transcriber) startTranscriberPool() {
+	slog.Debug("live captions: starting transcriber pool")
+
+	// Setup the transcribers
+	for i := 0; i < parallelTranscribersPool; i++ {
+		go t.handleTranscriptionRequests(i)
+	}
+}
+
+func (t *Transcriber) handleTranscriptionRequests(num int) {
+	slog.Debug(fmt.Sprintf("live captions: starting transcriber #%d", num))
+	t.captionWg.Add(1)
+
+	transcriber, err := t.newLiveCaptionsTranscriber()
+	if err != nil {
+		slog.Error("live captions: failed to create transcriber",
+			slog.String("err", err.Error()))
+		return
+	}
+	defer func() {
+		err := transcriber.Destroy()
+		if err != nil {
+			slog.Error("live captions: failed to destroy transcriber",
+				slog.String("err", err.Error()))
+		}
+		t.captionWg.Done()
+	}()
+
+	for {
+		select {
+		case <-t.captionDoneCh:
+			slog.Debug(fmt.Sprintf("live captions: closing transcriber #%d", num))
+			return
+		case packet := <-t.captionQueue:
+			// TODO: a context with expiry for this?
```

Comment:
Ideally yes but not a big deal to be honest as Transcribe should return eventually.